Implement AbortSignal.any()

cdumez opened the pull request:
#### b194103750b661e619135309652b481fe9704103
<pre>
Implement AbortSignal.any()
<a href="https://bugs.webkit.org/show_bug.cgi?id=256176">https://bugs.webkit.org/show_bug.cgi?id=256176</a>
rdar://109056627

Reviewed by Brent Fulgham.

Implement AbortSignal.any() as per:
- <a href="https://github.com/whatwg/dom/pull/1152">https://github.com/whatwg/dom/pull/1152</a>

The feature is behind a runtime flag, disabled by default.

* LayoutTests/imported/w3c/web-platform-tests/dom/abort/abort-signal-any.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/dom/abort/abort-signal-any.any.worker-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/bindings/js/JSAbortSignalCustom.cpp:
(WebCore::JSAbortSignalOwner::isReachableFromOpaqueRoots):
* Source/WebCore/dom/AbortSignal.cpp:
(WebCore::AbortSignal::any):
(WebCore::AbortSignal::addSourceSignal):
(WebCore::AbortSignal::addDependentSignal):
(WebCore::AbortSignal::signalAbort):
* Source/WebCore/dom/AbortSignal.h:
* Source/WebCore/dom/AbortSignal.idl:

Canonical link: <a href="https://commits.webkit.org/264163@main">https://commits.webkit.org/264163@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8c410bb1da167786771d994464e4e83596ea2586

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6898 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7128 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7311 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8497 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7145 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6903 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/8375 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7067 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/10045 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7020 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/7686 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6353 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8597 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/5067 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6263 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/14051 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/5822 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/6752 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6350 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/9156 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/6473 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6830 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/5600 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/7038 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6206 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1647 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1635 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10391 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/7229 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6584 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1783 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -84,6 +84,20 @@ AVFoundationEnabled:
+    WebKit:
+      default: false
+    WebCore:
+      default: false
```
How does it work that the tests pass? Are features always enabled for tests?

cdumez replied:
Yes, features get enabled for tests.

cdumez replied:
Since I marked it as “preview”, I believe it would get enabled for Safari Technology Preview as well.

brentfulgham replied:
This is correct. `testable` is the first stage where we activate features while running layout tests. `preview` takes it a step further, and activates for Safari Technology Preview.

annevk reviewed (commented):
Non-reviewer looks good.

I suppose we don't need an equivalent of "create a dependent abort signal" as callers will be able to invoke `AbortSignal::any` directly.

brentfulgham commented on LayoutTests/imported/w3c/web-platform-tests/dom/abort/abort-signal-any.any-expected.txt:
```diff
@@ -1,12 +1,12 @@
+PASS AbortSignal.any() signals are composable (using AbortController)
+PASS AbortSignal.any() works with signals returned by AbortSignal.timeout() (using AbortController)
+PASS AbortSignal.any() works with intermediate signals (using AbortController)
+PASS Abort events for AbortSignal.any() signals fire in the right order (using AbortController)
```
Lovely!

brentfulgham commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -84,6 +84,20 @@ AVFoundationEnabled:
+  category: dom
+  status: preview
+  humanReadableName: "AbortSignal.any() operation"
+  humanReadableDescription: "AbortSignal.any() operation"
```
The description is perhaps a bit terse...

cdumez replied:
It seemed clear enough to me. I don't think I would be able to explain what the operation does in a short enough sentence. Do you have a better proposal?

brentfulgham replied:
It would be more consistent to say:

"Enable AbortSignal.any() operation"

brentfulgham replied:
That's how the other features are defined, even though it doesn't provide very much more context -- at least it explains that the purpose of the toggle is to Enable the feature.

cdumez replied:
I added "Enable"

brentfulgham commented on Source/WebCore/dom/AbortSignal.cpp:
```diff
@@ -86,6 +103,25 @@ AbortSignal::AbortSignal(ScriptExecutionContext* context, Aborted aborted, JSC::
+        return;
+    }
+    ASSERT(!signal.aborted());
+    ASSERT(!signal.isDependent());
```
This ASSERT seems unnecessary, unless it's possible for a single to move from dependent to non-dependent; I don't see any asynchronous calls so it seems impossible for this to ever be `!isDependent()`

cdumez replied:
I can drop them, the reason I added them is that they were in the spec text.

brentfulgham replied:
If the specifications suggests asserting this precondition, I have no objection to leaving it in the code. It just seemed weird to branch on some value, then assert the value again.

cdumez replied:
I removed the `!signal.isDependent()` assertion. It is in the spec text but I agree it is obvious enough that it seems unnecessary.

annevk replied:
In the spec things are organized a bit differently such that this is not immediately obvious. Your code seems a bit more elegant due to the separation of "add source signals". Perhaps something to contribute upstream at some point.

brentfulgham reviewed (approved):
Looks great! r=me

cdumez committed:
Implement AbortSignal.any()
https://bugs.webkit.org/show_bug.cgi?id=256176
rdar://109056627

Reviewed by Brent Fulgham.

Implement AbortSignal.any() as per:
- https://github.com/whatwg/dom/pull/1152

The feature is behind a runtime flag, disabled by default.

* LayoutTests/imported/w3c/web-platform-tests/dom/abort/abort-signal-any.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/dom/abort/abort-signal-any.any.worker-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/bindings/js/JSAbortSignalCustom.cpp:
(WebCore::JSAbortSignalOwner::isReachableFromOpaqueRoots):
* Source/WebCore/dom/AbortSignal.cpp:
(WebCore::AbortSignal::any):
(WebCore::AbortSignal::addSourceSignal):
(WebCore::AbortSignal::addDependentSignal):
(WebCore::AbortSignal::signalAbort):
* Source/WebCore/dom/AbortSignal.h:
* Source/WebCore/dom/AbortSignal.idl:

Canonical link: https://commits.webkit.org/264163@main
